apps/block_scout_web/lib/block_scout_web/controllers/smart_contract_controller.ex

```diff
@@ -33,9 +36,18 @@ defmodule BlockScoutWeb.SmartContractController do
           end
         else
           if contract_type == "proxy" do
-            Reader.read_only_functions_proxy(address_hash, implementation_address_hash_string)
+            functions_proxy = Reader.read_only_functions_proxy(address_hash, implementation_address_hash_string)
+            if functions_proxy == [] do
+              Reader.read_only_functions(address_hash)
+            else
+              functions_proxy
+            end
           else
-            Reader.read_only_functions(address_hash)
+            if contract_type_from_address == :regular do
```

Comment:
Why do we need both variables `contract_type` and `contract_type_from_address`? Feels like checking if `contact_type == "proxy"` is not enough to say with confidence if a contract is a proxy or not, I think we should rather fix that than adding additional conditional logic here.

Reply:
The `contract_type` has the information about the tab that you are in, so in a proxy could be regular/proxy but in a normal contract is only regular because there is not tab for proxy. The `contract_type_from_address` is the variable that means that the address that you are visiting is really a contract proxy or a normal contract. So we need to double check to be sure that when we are in a proxy we are only showing the implementation address in the Read contract tab, but when the contract is regular it should be show all the methods of the contract

Reply:
I believe that the combination of two parameters `contract_type` and `action` unambiguously define the currently selected tab. This information should be enough to render the relevant form, no? You can't get to the `Read Proxy` page for a regular contract as this tab will be hidden.
Looking at the code, `Read Contract` should contain read-only functions of the implementation contract itself. `Read Proxy` should contain read-only functions of the proxy contract and so on. Please take a look at an example for xdai: https://blockscout.com/xdai/mainnet/address/0x75Df5AF045d91108662D8080fD1FEFAd6aA0bb59/read-contract